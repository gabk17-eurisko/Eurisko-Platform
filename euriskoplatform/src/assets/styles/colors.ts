export enum COLORS {
  black = '#000',
  white = '#fff',
  gray = '#bababa',
  grey = '#757575',
  error = '#c63a38',
  disabled = 'rgba(0,0,0,0.1)',
}
