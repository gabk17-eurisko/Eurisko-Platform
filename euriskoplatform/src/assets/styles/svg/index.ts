import Close from './Close.svg';

export {Close};
