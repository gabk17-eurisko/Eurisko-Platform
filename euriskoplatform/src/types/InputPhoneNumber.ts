import {ReactNode} from 'react';
import {StyleProp, TextStyle} from 'react-native';
import {IInputTextProps} from './InputText.types';

export enum LanguageKey {
  ar = 'ar',
  en = 'en',
}

type Language = keyof typeof LanguageKey;

interface IInputPhoneNumberProps extends IOmittedIInputPhoneNumberProps {
  /**
   * Extension placement on the start of the input
   *
   * @type {?boolean}
   * @default false
   */
  extensionPositionEnd?: boolean;

  /**
   * Boolean that shows country extension instead of the its flag
   *
   * @type {?boolean}
   * @default false
   */
  showExtension?: boolean;

  /**
   * Custom text for the modal title header
   *
   * @type {?string}
   * @default 'Select your country'
   */
  modalTitle?: string;

  /**
   * Custom text for the close modal pressable
   *
   * @type {?string}
   * @default 'Close'
   */
  modalCloseText?: string;

  /**
   * Custom styling for the text input
   *
   * @type {?TextStyle}
   * @default {}
   */
  inputTextStyle?: StyleProp<TextStyle>;

  /**
   * Custom styling for the extension text
   *
   * @type {?TextStyle}
   * @default {}
   */
  extensionTextStyle?: StyleProp<TextStyle>;

  /**
   * Custom svg for close pressable
   *
   * @type {?ReactNode}
   * @default null
   */
  modalCloseSVG?: ReactNode;

  /**
   * Reverse modal content layout
   *
   * @type {?boolean}
   * @default false
   */
  reverse?: boolean;

  /**
   * Country and extension language support
   *
   * @type {?Language}
   * @default "en"
   */
  language?: Language;

  /**
   * Function that gets extenstion value and input value from component
   *
   * @type {Function}
   */
  getInputValue: Function;
}

type IOmittedIInputPhoneNumberProps = Omit<IInputTextProps, 'keyboardType'>;

export {IInputPhoneNumberProps};
