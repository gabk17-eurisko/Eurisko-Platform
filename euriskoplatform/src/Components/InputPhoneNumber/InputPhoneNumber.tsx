import {
  View,
  Text,
  Modal,
  Pressable,
  FlatList,
  SafeAreaView,
  TextInput,
} from 'react-native';
import countriesLibrary from 'i18n-iso-countries';
import React, {useCallback, useEffect, useState} from 'react';
import {getCountryCallingCode} from 'libphonenumber-js';
import styles from '../../assets/styles';
import InputText from '../InputText/InputText';
// @ts-ignore
import Flag from 'react-native-flags';
import InputContainer from '../InputContainer/InputContainer';
import {
  IInputPhoneNumberProps,
  LanguageKey,
} from '../../types/InputPhoneNumber';

countriesLibrary.registerLocale(require('i18n-iso-countries/langs/en.json'));
countriesLibrary.registerLocale(require('i18n-iso-countries/langs/ar.json'));
const countries = countriesLibrary.getNames('en', {select: 'official'});

let countriesArray = [] as any;

for (let [key, value] of Object.entries(countries)) {
  try {
    countriesArray.push({
      alpha2: key,
      name: value,
      nameAr: countriesLibrary.getName(key, 'ar'),
      code: `+${getCountryCallingCode(key as any)}`,
      codeAr: `+${Number(getCountryCallingCode(key as any)).toLocaleString(
        'ar-EG',
      )}`,
    });
  } catch {}
}

const defaultProps = {
  extensionPositionStart: false,
  showExtension: false,
  modalTitle: 'Select your country',
  modalCloseText: 'Close',
  inputTextStyle: {},
  extenstionTextStyle: {},
  modalCloseSVG: null,
  reverse: false,
  language: LanguageKey.en,
};

const InputPhoneNumber = ({
  disabled,
  disabledStyle,
  errorMessage,
  errorStyle,
  inputStyle,
  label,
  labelStyle,
  showErrors,
  extensionPositionEnd,
  showExtension,
  modalTitle,
  modalCloseText,
  inputTextStyle,
  extensionTextStyle,
  modalCloseSVG,
  reverse,
  language,
  getInputValue,
  ...props
}: IInputPhoneNumberProps) => {
  const [inputValue, setInputValue] = useState<string>('');
  const [extValue, setExtValue] = useState<string>('');
  const [filteredCountries, setFilteredCountries] = useState<Array<Object>>([]);
  const [countryExtension, setCountryExtension] = useState<any>({
    alpha2: 'US',
    name: 'United States of America',
    nameAr: 'الولايات المتحدة',
    code: '+1',
    codeAr: '+١',
  });
  const [modalVisible, setModalVisible] = useState<boolean>(false);

  const isArabic = language === 'ar';

  const renderItem = useCallback(
    ({item}: any) => {
      return (
        <Pressable
          onPress={() => {
            setModalVisible(false);
            setFilteredCountries([]);
            setCountryExtension(item);
            setInputValue('');
          }}>
          <View style={styles.countryListItem}>
            {reverse ? (
              <>
                <View style={styles.countryListItemView}>
                  <Text style={styles.countryListItemText}>
                    {isArabic ? item?.nameAr : item?.name}
                  </Text>
                  <Flag size={32} code={item?.alpha2} />
                </View>
                <Text style={styles.itemCode}>
                  {isArabic ? item?.codeAr : item?.code}
                </Text>
              </>
            ) : (
              <>
                <Text style={styles.itemCode}>
                  {isArabic ? item?.codeAr : item?.code}
                </Text>
                <View style={styles.countryListItemView}>
                  <Flag size={32} code={item?.alpha2} />
                  <Text style={styles.countryListItemText}>
                    {isArabic ? item?.nameAr : item?.name}
                  </Text>
                </View>
              </>
            )}
          </View>
        </Pressable>
      );
    },
    [setCountryExtension, setModalVisible],
  );

  useEffect(() => {
    getInputValue(
      isArabic ? countryExtension?.codeAr : countryExtension?.code,
      inputValue,
    );
  }, [countryExtension, inputValue]);

  return (
    <>
      <InputContainer
        testID="container"
        disabled={disabled}
        disabledStyle={disabledStyle}
        errorMessage={errorMessage}
        errorStyle={errorStyle}
        inputStyle={[styles.inputContainerPhoneNumber, inputStyle]}
        label={label}
        labelStyle={labelStyle}
        showErrors={showErrors}>
        <View
          testID="extensionPosition"
          style={{
            flexDirection: extensionPositionEnd ? 'row' : 'row-reverse',
          }}>
          <TextInput
            {...props}
            textAlign={isArabic ? 'right' : 'left'}
            editable={!disabled}
            value={inputValue}
            keyboardType="number-pad"
            onChangeText={setInputValue}
            style={[styles.inputPhoneNumber, inputTextStyle]}
          />
          <Pressable
            testID="button"
            style={styles.inputPhoneNumberIconWrapper}
            onPress={() =>
              disabled ? null : setModalVisible((prev: boolean) => !prev)
            }>
            {showExtension ? (
              <Text
                testID="extension"
                style={[styles.extensionTxtPhoneNumber, extensionTextStyle]}>
                {isArabic ? countryExtension?.codeAr : countryExtension?.code}
              </Text>
            ) : (
              <Flag size={32} code={countryExtension?.alpha2 ?? 'US'} />
            )}
          </Pressable>
        </View>
      </InputContainer>

      <Modal
        animationType="slide"
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible((prev: boolean) => !prev);
        }}>
        <SafeAreaView style={styles.safeAreaPhoneNumber}>
          <View style={styles.wrapperPhoneNumber}>
            <View
              testID="modalHeader"
              style={[
                styles.titleViewPhoneNumber,
                reverse && {flexDirection: 'row-reverse'},
              ]}>
              <Text style={styles.titleTextPhoneNumber}>{modalTitle}</Text>
              <Pressable
                style={styles.closeButton}
                onPress={() => {
                  setModalVisible(false);
                  setFilteredCountries([]);
                  setInputValue('');
                }}>
                {modalCloseSVG ? (
                  modalCloseSVG
                ) : (
                  <Text style={styles.titleTextPhoneNumber}>
                    {modalCloseText}
                  </Text>
                )}
              </Pressable>
            </View>
            <View style={styles.modalInputContainerPhoneNumber}>
              <InputText
                testID="searchInput"
                value={extValue}
                textAlign={isArabic ? 'right' : 'left'}
                onChangeText={e => {
                  setExtValue(e);
                  setFilteredCountries(() => {
                    return countriesArray.filter(
                      (country: any) =>
                        country.name.toLowerCase().includes(e.toLowerCase()) ||
                        country.nameAr
                          .toLowerCase()
                          .includes(e.toLowerCase()) ||
                        country.code.includes(e) ||
                        country.codeAr.includes(e) ||
                        country.alpha2.includes(e.toUpperCase()),
                    );
                  });
                }}
              />
            </View>
            <FlatList
              testID="flatlist"
              showsVerticalScrollIndicator={false}
              keyboardShouldPersistTaps="always"
              style={[styles.flatListPhoneNumber]}
              data={
                filteredCountries.length > 0
                  ? filteredCountries
                  : countriesArray
              }
              renderItem={renderItem}
              keyExtractor={item => item?.alpha2}
            />
          </View>
        </SafeAreaView>
      </Modal>
    </>
  );
};

InputPhoneNumber.defaultProps = defaultProps;

InputPhoneNumber.displayName = 'InputPhoneNumber';

export default InputPhoneNumber;
