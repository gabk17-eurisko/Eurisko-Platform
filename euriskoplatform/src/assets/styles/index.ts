import {Platform, StyleSheet, SafeAreaView} from 'react-native';
import {COLORS} from './colors';

export default StyleSheet.create({
  inputContainerStyle: {
    flex: 1,
  },
  inputViewStyle: {
    borderWidth: 1,
    width: '100%',
    padding: 10,
    borderRadius: 5,
    borderColor: COLORS.gray,
  },
  inputLabelStyles: {
    marginBottom: 10,
    fontSize: 15,
    fontWeight: '400',
  },
  errorTextstyle: {
    color: COLORS.error,
    marginTop: 10,
    fontSize: 13,
    fontWeight: '300',
  },
  disabledInputStyle: {
    backgroundColor: COLORS.disabled,
  },
  errorInputStyle: {
    borderColor: COLORS.error,
  },
  passwordInput: {
    flex: 1,
  },
  passwordIconWrapper: {
    paddingHorizontal: 3,
  },
  passwordStrengthBarViewStyle: {
    height: 2,
    marginVertical: 5,
    borderRadius: 1,
  },
  countryListItem: {
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-between',
    height: 50,
    alignItems: 'center',
  },
  itemCode: {
    color: COLORS.grey,
  },
  countryListItemView: {
    flexDirection: 'row-reverse',
    alignItems: 'center',
  },
  countryListItemText: {
    paddingLeft: Platform.OS === 'android' ? 5 : 15,
    paddingRight: Platform.OS === 'ios' ? 5 : 15,
    color: COLORS.black,
  },
  wrapperPhoneNumber: {
    backgroundColor: COLORS.white,
    paddingHorizontal: 30,
    alignItems: 'center',
    elevation: 5,
  },
  titleViewPhoneNumber: {
    flexDirection: 'row',
    alignItems: 'center',
    width: '100%',
    marginBottom: 30,
    marginTop: 20,
  },
  titleTextPhoneNumber: {
    color: COLORS.black,
    fontSize: 17,
  },
  closeButton: {
    zIndex: 10,
    position: 'absolute',
    right: 0,
  },
  modalInputContainerPhoneNumber: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 20,
  },
  flatListPhoneNumber: {
    width: '100%',
    marginBottom: 120,
  },
  safeAreaPhoneNumber: {
    flex: 1,
  },
  inputPhoneNumber: {
    flex: 1,
    fontSize: 18,
  },
  inputPhoneNumberIconWrapper: {
    paddingHorizontal: 10,
  },
  extensionTxtPhoneNumber: {
    fontSize: 18,
  },
  inputContainerPhoneNumber: {
    height: 50,
    justifyContent: 'center',
  },
});
