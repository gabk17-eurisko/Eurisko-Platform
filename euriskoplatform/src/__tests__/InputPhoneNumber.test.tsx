import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react-native';
import InputPhoneNumber from '../Components/InputPhoneNumber/InputPhoneNumber';
// @ts-ignore
import renderer from 'react-test-renderer';
import {Close} from '../assets/styles/svg';
import '@testing-library/jest-native/extend-expect';

describe('<InputPhoneNumber/>', () => {
  const inputPhoneNumber = (
    <InputPhoneNumber
      getInputValue={(e: string, v: string) => {}}
      testID="input"
    />
  );

  const inputPhoneNumberWithProps = (
    <InputPhoneNumber
      getInputValue={(e: string, v: string) => {}}
      testID="input"
      extensionPositionEnd
      showExtension
      modalTitle="Pick"
      modalCloseText="Exit"
      inputTextStyle={{fontSize: 14}}
      extensionTextStyle={{fontSize: 20}}
      reverse
    />
  );

  const inputPhoneNumberArabic = (
    <InputPhoneNumber
      getInputValue={(e: string, v: string) => {}}
      language="ar"
      showExtension
    />
  );

  const inputPhoneNumberShowExtension = (
    <InputPhoneNumber
      getInputValue={(e: string, v: string) => {}}
      showExtension
    />
  );

  const inputPhoneNumberAsJson = renderer
    .create(
      <InputPhoneNumber
        getInputValue={(e: string, v: string) => {}}
        testID="input"
      />,
    )
    .toJSON();

  it('rendering successfully', () => {
    expect(inputPhoneNumber).toBeDefined();
  });

  it('renders correctly', () => {
    expect(inputPhoneNumberAsJson).toMatchSnapshot();
  });

  it('Outputs the correct text', () => {
    const {getByTestId} = render(inputPhoneNumber);

    const input = getByTestId('input');
    fireEvent.changeText(input, '123');
    expect(input.props.value).toBe('123');
  });

  it(`renders with the modal closed`, () => {
    const {queryAllByText} = render(inputPhoneNumber);
    expect(queryAllByText('Close').length).toBe(0);
  });

  it('fires the assigned event on press and opens the modal', () => {
    const {getByTestId, getByText} = render(inputPhoneNumber);
    const button = getByTestId('button');
    fireEvent.press(button);

    const isModalOpen = getByText('Select your country');
    expect(isModalOpen).toBeTruthy();
  });

  it('renders the component with prop extensionPositionStart', () => {
    const {getByTestId} = render(inputPhoneNumberWithProps);

    //this
    const extPosistionView = getByTestId('extensionPosition');
    expect(extPosistionView).toHaveStyle({flexDirection: 'row'});
    //is equivalent to that
    const flexDirection = extPosistionView.props.style.flexDirection;
    expect(flexDirection).toBe('row');
  });

  it('renders the component with styling props', () => {
    const {getByTestId} = render(inputPhoneNumberWithProps);

    const inputStyle = getByTestId('input');
    expect(inputStyle).toHaveStyle({fontSize: 14});

    const extensionStyle = getByTestId('extension');
    expect(extensionStyle).toHaveStyle({fontSize: 20});
  });

  it('renders the component with extension showing instead of flag', () => {
    const {getByText} = render(inputPhoneNumberWithProps);

    const extension = getByText('+1');
    expect(extension).toBeTruthy();
  });

  it('renders the component in arabic', () => {
    const {getByText} = render(inputPhoneNumberArabic);

    const extension = getByText('+١');
    expect(extension).toBeTruthy();
  });

  it('is possible to select country', () => {
    const {getByText, getByTestId, queryAllByText} = render(
      inputPhoneNumberShowExtension,
    );

    const button = getByTestId('button');
    fireEvent.press(button);

    const isModalOpen = getByText('Select your country');
    expect(isModalOpen).toBeTruthy();

    const country = getByText('+355');
    expect(country).toBeTruthy();

    fireEvent.press(country);

    const isModalClosed = queryAllByText('Select your country');
    expect(isModalClosed).toHaveLength(0);

    const extension = getByText('+355');
    expect(extension).toBeTruthy();
  });

  it('is possible to search for a certain extension and select it', () => {
    const {getByText, getByTestId, queryAllByText} = render(
      inputPhoneNumberShowExtension,
    );

    const arr = ['961', 'lebanon', 'Leba', 'Lb', 'لبنان'];

    const button = getByTestId('button');
    fireEvent.press(button);

    const isModalOpen = getByText('Select your country');
    expect(isModalOpen).toBeTruthy();

    const searchInput = getByTestId('searchInput');
    expect(searchInput).toBeTruthy();

    for (let i = 0; i < arr.length; i++) {
      fireEvent.changeText(searchInput, arr[i]);
      const searchedExtension = getByText('+961');
      expect(searchedExtension).toBeTruthy();
    }

    const searchedExtension = getByText('+961');
    expect(searchedExtension).toBeTruthy();

    fireEvent.press(searchedExtension);

    const isModalClosed = queryAllByText('Select your country');
    expect(isModalClosed).toHaveLength(0);

    const extension = getByText('+961');
    expect(extension).toBeTruthy();
  });
});
