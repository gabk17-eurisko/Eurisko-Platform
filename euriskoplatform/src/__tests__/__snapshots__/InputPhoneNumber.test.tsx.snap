// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputPhoneNumber/> renders correctly 1`] = `
[
  <View
    style={
      [
        null,
        {
          "borderColor": "#bababa",
          "borderRadius": 5,
          "borderWidth": 1,
          "padding": 10,
          "width": "100%",
        },
        [
          {
            "height": 50,
            "justifyContent": "center",
          },
          undefined,
        ],
        null,
      ]
    }
    testID="container"
  >
    <View
      style={
        {
          "flexDirection": "row-reverse",
        }
      }
      testID="extensionPosition"
    >
      <TextInput
        editable={true}
        extensionPositionStart={false}
        extenstionTextStyle={{}}
        keyboardType="number-pad"
        onChangeText={[Function]}
        style={
          [
            {
              "flex": 1,
              "fontSize": 18,
            },
            {},
          ]
        }
        testID="input"
        textAlign="left"
        value=""
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "paddingHorizontal": 10,
          }
        }
        testID="button"
      >
        <Image
          source={
            {
              "testUri": "../../react-native-flags/flags/shiny/32/US.png",
            }
          }
          style={
            [
              {
                "height": 32,
                "width": 32,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  />,
]
`;
